packages/mds-compliance-api/handlers/get-violation-periods.ts

```diff
@@ -0,0 +1,148 @@
+import { ComplianceServiceClient, ComplianceSnapshotDomainModel } from '@mds-core/mds-compliance-service'
+import db from '@mds-core/mds-db'
+import { providerName, providers } from '@mds-core/mds-providers'
+import { ApiRequestQuery } from '@mds-core/mds-api-server'
+import express from 'express'
+import { parseRequest } from '@mds-core/mds-api-helpers'
+import { Policy, Timestamp } from '@mds-core/mds-types'
+import { BadParamsError, isDefined, now, uuid } from '@mds-core/mds-utils'
+import { ComplianceAggregate, ComplianceApiRequest, ComplianceApiResponse, ComplianceViolationPeriod } from '../@types'
+
+export type ComplianceApiGetViolationPeriodsRequest = ComplianceApiRequest &
+  ApiRequestQuery<'start_time' | 'end_time' | 'provider_ids' | 'policy_ids'>
+
+export type ComplianceApiGetViolationPeriodsResponse = ComplianceApiResponse<{
+  start_time: Timestamp
+  end_time: Timestamp
+  results: ComplianceAggregate[]
+}>
+
+interface ComplianceAggregateMap {
+  [k: string]: {
+    complianceSnapshotViolationGroupings: ComplianceSnapshotDomainModel[][]
+    complianceSnapshots: null | ComplianceSnapshotDomainModel[]
+  }
+}
+
+async function convertComplianceSnapshotsArrayToComplianceViolationPeriod(
+  snapshots: ComplianceSnapshotDomainModel[]
+): Promise<ComplianceViolationPeriod> {
+  const compliance_array_response_id = uuid()
+  await ComplianceServiceClient.createComplianceArrayResponse({
+    compliance_array_response_id,
+    compliance_snapshot_ids: snapshots.map(s => s.compliance_snapshot_id),
+    provider_id: snapshots[0].provider_id
+  })
+  return {
+    start_time: snapshots[0].compliance_as_of,
+    end_time: snapshots[snapshots.length - 1].compliance_as_of,
```

Comment:
I think this is different than how I’d originally intended `end_time` to work. `end_time` should indicate when the period has ended, but as currently implemented this would make it seem like all violation periods have ended, even if they’re still ongoing. Under the current implementation, if you only have one snapshot within a period, this will result in both the `start_time` and `end_time` being the same, and that is fundamentally not a period, as there is a time interval of 0.

A violation period should be considered ongoing until we know that it has finished, and we know that it has finished once we’ve seen a snapshot with a `violation: 0`, so the timestamp for the `end_time` should come from the snapshot that brings it back into compliance, not from the last snapshot that we knew for sure was in violation. For violations that are still ongoing, that snapshot won’t exist at the time of the violation periods being fetched, so you can just simply put in `end_time: null`. This makes it clear to the client that a violation is ongoing.

Reply:
👍